File: app/src/main/java/com/niedzwiecki/przemyslguide/ui/base/BaseViewModel.java
package com.niedzwiecki.przemyslguide.ui.base;

import android.databinding.BaseObservable;
import android.os.Bundle;

/**
 * Created by niedzwiedz on 29.06.17.
 */

public class BaseViewModel<T extends Navigator> extends BaseObservable implements ViewModel {

    Navigator navigator;
    private T context;

    @Override
    public void attachNavigator(Navigator navigator) {
        this.navigator = navigator;
    }

    @Override
    public void detachNavigator() {
        navigator = null;
    }

    @Override
    public void onActivityResult(Navigator navigator, int requestCode, int resultCode, Bundle data) {

    }

    @Override
    public void onDestroy() {

    }

    @Override
    public void onCreate() {

    }

    @Override
    public void saveInstanceState(Bundle bundle) {

    }

    @Override
    public void restoreInstanceState(Bundle bundle) {

    }

    @Override
    public void attachContext(Navigator context) {
        this.context = (T) context;
    }


    public boolean isNavigatorAttached() {
        return navigator != null;
    }

    public Navigator getNavigator() {
        checkNavigatorAttached();
        return navigator;
    }

    public void checkNavigatorAttached() {
        if (!isNavigatorAttached()) throw new MvpViewNotAttachedException();
    }

    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults, BaseActivity activity) {

    }

    public static class MvpViewNotAttachedException extends RuntimeException {

        public MvpViewNotAttachedException() {
            super("Please call attachNavigator(Context) before" +
                    " requesting data to the ViewModel");
        }
    }

}

File: app/src/main/java/com/niedzwiecki/przemyslguide/ui/base/BaseViewModel.kt
package com.niedzwiecki.przemyslguide.ui.base

import android.databinding.BaseObservable
import android.os.Bundle

/**
 * Created by niedzwiedz on 29.06.17.
 */

open class BaseViewModel<T : Navigator> : BaseObservable(), ViewModel {

    internal var navigator: Navigator? = null
    private var context: T? = null

    val isNavigatorAttached: Boolean
        get() = navigator != null

    override fun attachNavigator(navigator: Navigator) {
        this.navigator = navigator
    }

    override fun detachNavigator() {
        navigator = null
    }

    override fun onActivityResult(navigator: Navigator, requestCode: Int, resultCode: Int, data: Bundle) {

    }

    override fun onDestroy() {

    }

    override fun onCreate() {

    }

    override fun saveInstanceState(bundle: Bundle) {

    }

    override fun restoreInstanceState(bundle: Bundle) {

    }

    override fun attachContext(context: Navigator) {
        this.context = context as T
    }

    fun getNavigator(): Navigator? {
        checkNavigatorAttached()
        return navigator
    }

    fun checkNavigatorAttached() {
        if (!isNavigatorAttached) throw MvpViewNotAttachedException()
    }

    fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray, activity: BaseActivity) {

    }

    class MvpViewNotAttachedException : RuntimeException("Please call attachNavigator(Context) before" + " requesting data to the ViewModel")

}
